oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizationContext.java

```diff
@@ -0,0 +1,272 @@
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A context that holds authorization-specific state and is used by an {@link OAuth2AuthorizedClientProvider}
+ * when attempting to authorize (or re-authorize) an OAuth 2.0 Client.
+ *
+ * @author Joe Grandja
+ * @since 5.2
+ * @see OAuth2AuthorizedClientProvider
+ */
+public final class OAuth2AuthorizationContext {
+	private ClientRegistration clientRegistration;
+	private Authentication principal;
+	private OAuth2AuthorizedClient authorizedClient;
+	private Map<String, Object> attributes;
+
+	private OAuth2AuthorizationContext() {
+	}
+
+	/**
+	 * Returns the {@link ClientRegistration client} requiring authorization.
+	 *
+	 * @return the {@link ClientRegistration}
+	 */
+	public ClientRegistration getClientRegistration() {
+		return this.clientRegistration;
+	}
+
+	/**
+	 * Returns the {@code Principal} to be associated with the authorized client.
+	 *
+	 * @return the {@code Principal} to be associated with the authorized client
+	 */
+	public Authentication getPrincipal() {
+		return this.principal;
+	}
+
+	/**
+	 * Returns the {@link OAuth2AuthorizedClient authorized client} which requires re-authorization
+	 * or {@code null} if the {@link #getClientRegistration() client} needs to be authorized.
+	 *
+	 * @return the {@link OAuth2AuthorizedClient} which requires re-authorization or {@code null} if the client needs to be authorized
+	 */
+	@Nullable
+	public OAuth2AuthorizedClient getAuthorizedClient() {
+		return this.authorizedClient;
+	}
+
+	/**
+	 * Returns the attributes associated to the context.
+	 *
+	 * @return a {@code Map} of the attributes associated to the context
+	 */
+	public Map<String, Object> getAttributes() {
+		return this.attributes;
+	}
+
+	/**
+	 * Returns the value of an attribute associated to the context, or {@code null} if not available.
+	 *
+	 * @param name the name of the attribute
+	 * @param <T> the type of the attribute
+	 * @return the value of the attribute associated to the context
+	 */
+	@Nullable
+	@SuppressWarnings("unchecked")
+	public <T> T getAttribute(String name) {
+		return (T) this.getAttributes().get(name);
+	}
+
+	/**
+	 * Returns {@code true} if the client needs to be authorized, otherwise {@code false}.
+	 *
+	 * @return {@code true} if the client needs to be authorized, otherwise {@code false}.
+	 */
+	public boolean authorizationRequired() {
+		return getAuthorizedClient() == null;
+	}
+
+	/**
+	 * Returns {@code true} if the client needs to be re-authorized, otherwise {@code false}.
+	 *
+	 * @return {@code true} if the client needs to be re-authorized, otherwise {@code false}.
+	 */
+	public boolean reauthorizationRequired() {
+		return getAuthorizedClient() != null;
```

Comment:
Can you explain your reasoning a bit here? "Reauthorization is required if we have an authorized client" doesn't seem logical.

Actually, given the name, I'd expect logic like what `hasTokenExpired` has.

Would it make sense to either 1. actually check if reauthorization is indeed required or 2. rename the method to `reauthorizationRequested`? 

If it's not ideal to place the expiration logic here in the context, then I think that `reauthorizationRequested` aligns better with the application's intent when calling the `reauthorize` static method, i.e. the application doesn't actually care whether reauthorization truly is required, it wants it reauthorized either way.

I think it would be easy to make this mistake, say if I were building a Zuul filter, or similar:

```java
OAuth2AuthorizationContext.Builder contextBuilder = reauthorize(authorizedClient);
if (contextBuilder.reauthorizeRequired()) {
    // ... application assumes that this means the token is expired
}
```

Reply:
I second this sentiment. More particularly, I'm not sure that it makes sense to embed logic into a domain object.

I'm also curious what defines re-authorization. Does this indicate that authorization has expired, or is it more inclusive (i.e. if the client just wants to request a new access token).

I wonder if it makes more sense to try and update the APIs be explicit and align with the OAuth APIs. OAuth2AuthorizedClientProvider would have two methods to get an access token and another to refresh it.

Reply:
@jzheaux @rwinch Let me explain the meaning behind reauthorization. A client may need to be reauthorized in the following cases:

1. access token is expired and an `OAuth2AuthorizedClient.refreshToken` is available, e.g. `authorization_code` or `password` client. NOTE: `client_credentials` client does not receive a refresh token as per spec recommendation.
2. access token is expired and an `OAuth2AuthorizedClient.refreshToken` is NOT available, e.g. `client_credentials` client do not have refresh token so this is a renewal (not a refresh). Furthermore, this case may apply to a `authorization_code` or `password` client if a refresh token is not available depending on the client and/or provider configuration.

It's important to note that a refresh token request can only happen when there is a refresh token available and the `refresh_token` grant is performed. In all other cases, it's getting a new access token OR renewing an expired access token for reauthorization cases.

> Would it make sense to either 1. actually check if reauthorization is indeed required or 2. rename the method to reauthorizationRequested?

It doesn't make sense to contain the access token expiration logic in the context. The expiration logic should be contained within the http client as it is today with `ServletOAuth2AuthorizedClientExchangeFilterFunction`. This is where the user would configure the `accessTokenExpiresSkew` as well. The http client should determine if the access token is expired or not and build the `OAuth2AuthorizationContext` for authorization or reauthorization.

> 2. rename the method to reauthorizationRequested?

I think this name makes a lot of sense. Let me propose the new names based on this?

```java
requestAuthorization(ClientRegistration clientRegistration)
boolean authorizationRequested()

requestReauthorization(OAuth2AuthorizedClient authorizedClient)
boolean reauthorizationRequested()
```

Or possibly an alternative name for `reauthorization`:

```java
requestRenewal(OAuth2AuthorizedClient authorizedClient)
boolean renewalRequested()
```

@jzheaux @rwinch What are your thoughts on these names?

Reply:
@rwinch 

> I wonder if it makes more sense to try and update the APIs be explicit and align with the OAuth APIs. OAuth2AuthorizedClientProvider would have two methods to get an access token and another to refresh it.

I don't think it makes sense to add 2 operations (e.g. `authorize` and `refresh`) to `OAuth2AuthorizedClientProvider`. The `refresh()` can only happen in the `refresh_token` case - `RefreshTokenOAuth2AuthorizedClientProvider`. Even if we renamed the operation to something more generic like `renew()`, not all `OAuth2AuthorizedClientProvider` would implement this operation (and vice-versa) so there would be `null` implementations. For example, `authorization_code` `OAuth2AuthorizedClientProvider` would only implement `authorize()` (not `renew()`) and `refresh_token` `OAuth2AuthorizedClientProvider` would implement `renew()` but never `authorize()`.

My preference would be to keep the one `authorize()` operation and it's up to the implementation to determine if the request is for authorization or reauthorization depending on what's in the context.

Reply:
Thanks for the explanation on not embedding the logic in the context object - I agree that it would be better to rename the methods.

I think what I'd do is:

```java
static Builder withAuthorizedClient(OAuth2AuthorizedClient authorizedClient)
static Builder withClientRegistration(ClientRegistration clientRegistration)

boolean authorizationRequested()
boolean reauthorizationRequested()
```

I worry that some semantics might be lost, but it's nice that it aligns with the team standard for other builders. The semantic distance could be nice, too, in case our definition of what it means to request reauthorization changes.

If that's not a concern, then I like:

```java
static Builder forReauthorization(OAuth2AuthorizedClient authorizedClient)
static Builder forAuthorization(ClientRegistration clientRegistration)
```

That would create `OAuth2AuthorizationContext.forReauthorization(...)`, which I think reads nicely.

I could buy `requestReauthorization` because of the symmetry with `reauthorizationRequested`, but I lean away from it since we are working with a builder. It seems to me that the only verb method we'd want for a builder would be `build()` itself.

Reply:
I like `forReauthorization()` and `forAuthorization()` the best. I'm going to go with this for now for next review.